src/queue/multilevel.rs

```diff
@@ -0,0 +1,834 @@
+        let mut rng = thread_rng();
+        let mut need_retry = true;
+        while need_retry {
+            need_retry = false;
+            let expected_level =
+                if rng.gen_ratio(self.manager.get_level0_chance(), u32::max_value()) {
+                    0
+                } else {
+                    (1..LEVEL_NUM - 1)
+                        .find(|_| rng.gen_ratio(CHANCE_RATIO, CHANCE_RATIO + 1))
+                        .unwrap_or(LEVEL_NUM - 1)
+                };
+            match self.level_injectors[expected_level].steal_batch_and_pop(&self.local_queue) {
+                Steal::Success(t) => return Some(into_pop(t, false)),
+                Steal::Retry => need_retry = true,
+                _ => {}
+            }
+            if !self.stealers.is_empty() {
+                let mut found = None;
+                for (idx, stealer) in self.stealers.iter().enumerate() {
+                    match stealer.steal_batch_and_pop(&self.local_queue) {
+                        Steal::Success(t) => {
+                            found = Some((idx, into_pop(t, false)));
+                            break;
+                        }
+                        Steal::Retry => need_retry = true,
+                        _ => {}
+                    }
+                }
+                if let Some((idx, task)) = found {
+                    let last_pos = self.stealers.len() - 1;
+                    self.stealers.swap(idx, last_pos);
+                    return Some(task);
+                }
+            }
+            for injector in self
+                .level_injectors
+                .iter()
+                .chain(&*self.level_injectors)
+                .skip(expected_level + 1)
+                .take(LEVEL_NUM - 1)
+            {
+                match injector.steal_batch_and_pop(&self.local_queue) {
+                    Steal::Success(t) => return Some(into_pop(t, false)),
+                    Steal::Retry => need_retry = true,
+                    _ => {}
+                }
+            }
+        }
+        None
+    }
+}
+
+/// The runner builder for multilevel task queues.
+///
+/// It can be created by [`Builder::runner_builder`].
+pub struct MultilevelRunnerBuilder<B> {
+    inner: B,
+    manager: Arc<LevelManager>,
+}
+
+impl<B, R, T> RunnerBuilder for MultilevelRunnerBuilder<B>
+where
+    B: RunnerBuilder<Runner = R>,
+    R: Runner<TaskCell = T>,
+    T: TaskCell,
+{
+    type Runner = MultilevelRunner<R>;
+
+    fn build(&mut self) -> Self::Runner {
+        MultilevelRunner {
+            inner: self.inner.build(),
+            manager: self.manager.clone(),
+        }
+    }
+}
+
+/// The runner for multilevel task queues.
+///
+/// The runner helps multilevel task queues collect additional information.
+/// [`MultilevelRunnerBuilder`] is the [`RunnerBuilder`] for this runner.
+pub struct MultilevelRunner<R> {
+    inner: R,
+    manager: Arc<LevelManager>,
+}
+
+impl<R, T> Runner for MultilevelRunner<R>
```

Comment:
Should be `MultiLevelRunner`.

Reply:
Here I suppose "multilevel" is a single word. If you prefer "multi-level", we should change all occurances to "MultiLevel" (for camel case) or "multi_level" (for snake case).

Reply:
Oh, I didn't realize it was a word.